cpp/arcticdb/storage/mongo/mongo_mock_client.cpp

```diff
@@ -0,0 +1,201 @@
+#include <bsoncxx/builder/basic/document.hpp>
+#include <mongocxx/exception/operation_exception.hpp>
+#include <mongocxx/exception/bulk_write_exception.hpp>
+#include <mongocxx/exception/query_exception.hpp>
+
+
+namespace arcticdb::storage::mongo {
+
+std::string MockMongoClient::get_failure_trigger(
+        const std::string& key,
+        StorageOperation operation_to_fail,
+        MongoError error_code) {
+    return fmt::format("{}#Failure_{}_{}", key, operation_to_string(operation_to_fail), static_cast<int>(error_code));
+}
+
+template <typename exception_type>
+MongoFailure create_failure(const std::string& message, MongoError error_code) {
+    static_assert(std::is_base_of<mongocxx::operation_exception, exception_type>::value, "exception_type must be a subclass of mongocxx::operation_exception");
+    if(error_code == MongoError::NoAcknowledge) return {std::monostate()};
+
+    bsoncxx::document::value empty_doc_value = bsoncxx::builder::basic::document{}.extract();
+    auto ec = std::error_code((int)error_code, std::generic_category());
+
+    return {exception_type(ec, std::move(empty_doc_value), message)};
+}
+
+MongoFailure get_failure(const std::string& message, StorageOperation operation, MongoError error_code) {
+    switch(operation) {
+        case StorageOperation::READ:
+            return create_failure<mongocxx::query_exception>(message, error_code);
+        case StorageOperation::WRITE:
+            return create_failure<mongocxx::bulk_write_exception>(message, error_code);
+        case StorageOperation::EXISTS:
+            return create_failure<mongocxx::query_exception>(message, error_code);
+        case StorageOperation::DELETE:
+            return create_failure<mongocxx::bulk_write_exception>(message, error_code);
+        case StorageOperation::DELETE_LOCAL:
+            return create_failure<mongocxx::bulk_write_exception>(message, error_code);
+        case StorageOperation::LIST:
+            return create_failure<mongocxx::query_exception>(message, error_code);
+        default:
+            util::raise_rte("Unknown operation used for error trigger");
+    }
+}
+
+std::optional<MongoFailure> MockMongoClient::has_failure_trigger(
+        const mongo_document_key& key,
+        StorageOperation operation) const {
+    auto key_id = get<2>(key);
+    auto failure_string_for_operation = "#Failure_" + operation_to_string(operation) + "_";
+    auto position = key_id.rfind(failure_string_for_operation);
+    if (position == std::string::npos) return std::nullopt;
+
+    try {
+        auto start = position + failure_string_for_operation.size();
+        auto error_code = MongoError(stoi(key_id.substr(start)));
+        auto error_message = fmt::format("Simulated Error, message: operation {}, error code {}",
+                                         operation_to_string(operation), static_cast<int>(error_code));
+
+        return get_failure(error_message, operation, error_code);
+    } catch (std::exception&) {
```

Comment:
What case does this catch cover? Do we really want to swallow these?

Reply:
This could happen if someone messes up with the special `#Failure_` string instead of using `get_failure()`. In that case for example, `stoi` could fail and throw an exception. We just swallow the exception and return no failure because it hasn't been generated using `get_failure()`.  This was also done for s3 and [azure mock clients](https://github.com/man-group/ArcticDB/pull/1331/files#diff-5cc7ef4386f48dc47564ce4f98e1ddd8aff872a4d37ac11ac6c4c9388a64ff73R55).

Note that potential exceptions returned by `get_failure` aren't swallowed because they are only thrown by the caller. This function just returns them and only swallows exceptions thrown during parsing of the error codes.

Reply:
I would think we want to blow up loudly is someone messes up the special string?

Reply:
As discussed this is exceptions as control flow - we expect the throw and catch for normal symbols that aren't the special failure strings. I think it's fair to say that fixing that here is scope creep, so I'm fine with this as is.